graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/tracing/FederatedTracingInstrumentation.java

```diff
@@ -0,0 +1,282 @@
+    private static final String KEY = "ftv1";
+    private static final String DEBUG_KEY = "ftv1_debug";
+
+    private boolean debuggingEnabled = false;
+
+    public FederatedTracingInstrumentation() {
+    }
+
+    public FederatedTracingInstrumentation(boolean debug) {
+        debuggingEnabled = debug;
+    }
+
+    @Override
+    public InstrumentationState createState() {
+        return new FederatedTracingState();
+    }
+
+    @Override
+    public CompletableFuture<ExecutionResult> instrumentExecutionResult(ExecutionResult executionResult, InstrumentationExecutionParameters parameters) {
+        FederatedTracingState state = parameters.getInstrumentationState();
+
+        Map<Object, Object> extensions = executionResult.getExtensions();
+        Map<Object, Object> tracingMap = new LinkedHashMap<>(extensions == null ? Collections.emptyMap() : extensions);
+
+        Reports.Trace trace = state.toProto();
+
+        byte[] bytes = trace.toByteArray();
+        tracingMap.put(KEY, Base64.getEncoder().encodeToString(bytes));
+
+        if (debuggingEnabled) {
+            tracingMap.put(DEBUG_KEY, trace.toString());
+        }
+
+        return CompletableFuture.completedFuture(new ExecutionResultImpl(executionResult.getData(), executionResult.getErrors(), tracingMap));
+    }
+
+    @Override
+    public InstrumentationContext<Object> beginFieldFetch(InstrumentationFieldFetchParameters parameters) {
+        FederatedTracingState state = parameters.getInstrumentationState();
+
+        long startNanos = System.nanoTime();
+        return whenCompleted((result, throwable) -> {
+            long endNanos = System.nanoTime();
+
+            DataFetchingEnvironment env = parameters.getEnvironment();
+            ExecutionStepInfo executionStepInfo = env.getExecutionStepInfo();
+
+            Node node = state.addNode(executionStepInfo, startNanos, endNanos);
+
+            if (throwable != null) {
+                node.addError(throwable);
+            }
+
+            if (result instanceof DataFetcherResult && ((DataFetcherResult) result).hasErrors()) {
+                DataFetcherResult theResult = (DataFetcherResult) result;
+                List<GraphQLError> errors = theResult.getErrors();
+                errors.forEach(node::addError);
+            }
+        });
+    }
+
+    /**
+     * Stores timing information and a map of nodes for field information.
+     */
+    private class FederatedTracingState implements InstrumentationState {
+        private final Instant startRequestTime;
+        private final long startRequestNanos;
+        private final LinkedHashMap<ExecutionPath, Node> nodeMap;
+
+        private FederatedTracingState() {
+            // record start time when creating instrumentation state for a request
+            startRequestTime = Instant.now();
+            startRequestNanos = System.nanoTime();
+
+            // Create node map with the root node
+            nodeMap = new LinkedHashMap<>();
```

Comment:
naming suggestion: `nodesByPath`?

Reply:
Also it seems like this state could likely be largely simplified to be built using the tree from `TraceNode.newBuilder`. I think that this object also means that building the node tree is a more arduous task at the end, and it would likely be a quick performance win (and code cleanliness win) to build the object directly rather than through a wrapper.